test/test_go.rb

```diff
@@ -1748,120 +1725,111 @@ def test_ctrl_t
+    tmux.until { |lines| lines[-2].end_with? '/foo' }
+  ensure
+    Dir.rmdir 'foo'
   end
 
   def test_alt_c_command
     set_var 'FZF_ALT_C_COMMAND', 'echo /tmp'
 
     tmux.prepare
-    tmux.send_keys 'cd /', :Enter
-
-    retries do
-      tmux.prepare
-      tmux.send_keys :Escape, :c
-      tmux.until { |lines| lines.item_count == 1 }
-    end
+    tmux.send_keys 'M-c'
+    tmux.until { |lines| lines[-1] == '> /tmp' }
     tmux.send_keys :Enter
 
     tmux.prepare
     tmux.send_keys :pwd, :Enter
-    tmux.until { |lines| lines[-1].end_with? '/tmp' }
+    tmux.until { |lines| lines[-2] == '/tmp' }
   end
 
   def test_ctrl_r
     tmux.prepare
-    tmux.send_keys 'echo 1st', :Enter; tmux.prepare
-    tmux.send_keys 'echo 2nd', :Enter; tmux.prepare
-    tmux.send_keys 'echo 3d',  :Enter; tmux.prepare
-    tmux.send_keys 'echo 3rd', :Enter; tmux.prepare
+    tmux.send_keys 'echo 1st', :Enter
+    tmux.prepare
+    tmux.send_keys 'echo 2nd', :Enter
+    tmux.prepare
+    tmux.send_keys 'echo 3d', :Enter
+    tmux.prepare
+    tmux.send_keys 'echo 3rd', :Enter
+    tmux.prepare
     tmux.send_keys 'echo 4th', :Enter
-    retries do
-      tmux.prepare
-      tmux.send_keys 'C-r'
-      tmux.until { |lines| lines.match_count.positive? }
-    end
+    tmux.prepare
     tmux.send_keys 'C-r'
+    tmux.until { |lines| lines[-1] == '>' }
     tmux.send_keys '3d'
-    tmux.until { |lines| lines[-3].end_with? 'echo 3rd' }
+    tmux.until { |lines| lines[-3].end_with? 'echo 3d' }
     tmux.send_keys :Enter
-    tmux.until { |lines| lines[-1] == 'echo 3rd' }
+    tmux.until { |lines| lines[-1] == '$ echo 3d' }
     tmux.send_keys :Enter
-    tmux.until { |lines| lines[-1] == '3rd' }
+    tmux.until { |lines| lines[-2] == '3d' }
   end
 
   def test_ctrl_r_multiline
+    tmux.prepare
     tmux.send_keys 'echo "foo', :Enter, 'bar"', :Enter
-    tmux.until { |lines| lines[-2..-1] == %w[foo bar] }
+    tmux.until { |lines| lines[-3..-2] == %w[foo bar] }
     tmux.send_keys 'C-r'
     tmux.until { |lines| lines[-1] == '>' }
     tmux.send_keys 'foo bar'
     tmux.until { |lines| lines[-3].end_with? 'bar"' }
     tmux.send_keys :Enter
     tmux.until { |lines| lines[-1].end_with? 'bar"' }
     tmux.send_keys :Enter
-    tmux.until { |lines| lines[-2..-1] == %w[foo bar] }
+    tmux.until { |lines| lines[-3..-2] == %w[foo bar] }
   end
 
   def test_ctrl_r_abort
     skip "doesn't restore the original line when search is aborted pre Bash 4" if shell == :bash && /(?<= version )\d+/.match(`#{Shell.bash} --version`).to_s.to_i < 4
     %w[foo ' "].each do |query|
       tmux.prepare
       tmux.send_keys(query)
-      tmux.until { |lines| lines[-1].start_with? query }
+      tmux.until { |lines| lines[-1].start_with? "$ #{query}" }
+      # Keys are possibly echoed before the (fish) shell is ready for
+      # C-r? In this case there's no other feedback to check for so add
+      # an arbitrary delay.
+      sleep 0.5 if shell == :fish
```

Comment:
I'd like to avoid this arbitrary delay. This seems to the case where the old Left, Right hack should work well. What do you think?

```sh
tmux.send_keys(query, :Left, :Right)
# $ foo^[[D^[[C (not ready)
# $ foo (ready)
tmux.until { |lines| lines[-1] == "$ #{query}" }
```

Reply:
I like it! But I tried it out and the intermittent failure returned: https://travis-ci.org/jablko/fzf/jobs/657850503#L534

Reply:
Is the link correct?

![image](https://user-images.githubusercontent.com/700826/75855521-d18a4980-5e35-11ea-8eeb-d30ec9cba119.png)

Reply:
Yes, the backtrace shows that it errored at ``test/test_go.rb:1842:in `block in test_ctrl_r_abort'``, which is: https://github.com/jablko/fzf/blob/28a39ddbc5a8761f151d8b05e3da4b92f98ad812/test/test_go.rb#L1842

So it sends `query` (in this case a double quote) followed by <kbd>:Left</kbd> and <kbd>:Right</kbd>. It finds `$ #{query}` (in this case `$ "`) so determines that the shell should be ready! It sends <kbd>C-r</kbd> but fish still stumbles and doesn't launch fzf :frowning_man:

Reply:
That's strange. Let me see.

Reply:
You should use the equality condition (`lines[-1] == "$ #{query}"`) [instead of `start_with?`](https://github.com/jablko/fzf/blob/28a39ddbc5a8761f151d8b05e3da4b92f98ad812/test/test_go.rb#L1840) which should return true before fish processes left and right arrow keys.

Reply:
If this left-right dance turns out to work well, we probably should put it in `tmux.prepare`.

Reply:
:man_facepalming: Sorry I missed `.start_with?`. Thanks for your patience! However there are still intermittent failures:
- - https://travis-ci.org/junegunn/fzf/jobs/658740657#L635
  - https://github.com/junegunn/fzf/blob/191c308a3dd00b83c2b088304521082a0c6b6d9a/test/test_go.rb#L1793
- - https://travis-ci.org/jablko/fzf/jobs/658775012#L527
  - https://github.com/jablko/fzf/blob/5cdaf389648d2415d915bd7408ff2c18ebf64356/test/test_go.rb#L1793
- - https://travis-ci.org/jablko/fzf/jobs/658781294#L524
  - https://github.com/jablko/fzf/blob/51fa9973434ebf990f9e9afd692d4283871e25a5/test/test_go.rb#L1793

I used `lines[-1] != "$ #{query}^[[D^[[C"` or `!lines[-1].start_with?("$ #{query}^")` instead of `lines[-1] == "$ #{query}"` because fish sometimes offers an autocompletion: https://travis-ci.org/jablko/fzf/jobs/658726395#L628

Reply:
> because fish sometimes offers an autocompletion

Hmm, this is painful. One thing I noticed on my machine is that I append a space after double-quote, the suggestion text disappears. So,

```ruby
tmux.send_keys(query, :Space, :Left, :Right)
```

might simplify the check. But anyway, this trick doesn't seem to completely prevent fish from not responding on CTRL-R.

There are other test cases for CTRL-R, should we apply the same hack on those cases as well? 

We used to wrap these key bindings with `retries` method,

- https://github.com/junegunn/fzf/blob/d9c6a0305b8c7f83a21fa583ae662660b940b1b6/test/test_go.rb#L1758-L1762
- https://github.com/junegunn/fzf/blob/d9c6a0305b8c7f83a21fa583ae662660b940b1b6/test/test_go.rb#L1839-L1843
- https://github.com/junegunn/fzf/blob/d9c6a0305b8c7f83a21fa583ae662660b940b1b6/test/test_go.rb#L1820-L1824

but you removed them all in this pull request. Maybe we should keep them?

Reply:
I think in those cases there's a prompt we can rely on instead?